src/core/service.c

```diff
@@ -2804,6 +2807,63 @@ static void service_run_next_main(Service *s) {
         (void) service_set_main_pidref(s, TAKE_PIDREF(pidref), &s->main_command->exec_status.start_timestamp);
 }
 
+static bool service_should_reload_extensions(Unit *u) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+        int r;
+
+        /* Only support this for notify-reload service types. */
+        if (s->type != SERVICE_NOTIFY_RELOAD)
+                return false;
+
+        /* TODO: Add support for user services, which can use
+         * ExtensionDirectories= + notify-reload. For now, skip for user
+         * services. */
+        if (u->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_unit_debug(UNIT(s), "Not reloading extensions for user services.");
+                return false;
+        }
+
+        r = exec_context_has_vpicked_extensions(&s->exec_context);
+        if (r < 0) {
+                log_unit_warning_errno(u, r, "Failed to determine if service should reload extensions, assuming false: %m");
+                return false;
+        }
+        return r > 0;
+}
+
+static int service_reload_extensions(Unit *u) {
+        /* TODO: do this asynchronously */
+        Service *s = ASSERT_PTR(SERVICE(u));
+        _cleanup_free_ char *propagate_dir = NULL;
+
+        if (!service_should_reload_extensions(u)) {
+                return 0;
+        }
+
+        propagate_dir = path_join("/run/systemd/propagate/", u->id);
+        if (!propagate_dir)
+                return -ENOMEM;
+
+        NamespaceParameters p = {
+                .private_namespace_dir = "/run/systemd",
+                .incoming_dir = "/run/systemd/incoming",
+                .propagate_dir = propagate_dir,
+                .runtime_scope = RUNTIME_SCOPE_SYSTEM,
```

Comment:
even if you don't support this for non-system services yet, please still initialize this properly from `m->runtime_scope`